tensorflow/lite/experimental/shlo/ops/dot_general.cc

```diff
@@ -0,0 +1,441 @@
+    if ((std::count(rhsb, rhsb + rhsb_size, i) == 0) &&
+        (std::count(rhsc, rhsc + rhsc_size, i) == 0)) {
+      rhs_result_dims.push_back(i);
+    }
+  }
+  for (size_t i = 0; i < lhs_result_dims.size(); ++i) {
+    output_shape_check.push_back(lhs.shape().Dim(lhs_result_dims[i]));
+  }
+  for (size_t i = 0; i < rhs_result_dims.size(); ++i) {
+    output_shape_check.push_back(rhs.shape().Dim(rhs_result_dims[i]));
+  }
+  for (size_t i = 0; i < output_rank; ++i) {
+    if (output.shape().Dim(i) != output_shape_check[i]) {
+      return absl::FailedPreconditionError(
+          "stablehlo.dot_general: Invalid output shape.");
+    }
+  }
+  if (lhs.IsPerAxisQuantized()) {
+    return absl::FailedPreconditionError(
+        "stablehlo.dot_general: The lhs tensor cannot be per-axis quantized.");
+  }
+  if (!lhs.IsPerTensorQuantized() && !rhs.IsQuantized() &&
+      lhs.tensor_element_type() != rhs.tensor_element_type()) {
+    return absl::FailedPreconditionError(
+        "stablehlo.dot_general: For non-quantized tensors the element type of "
+        "lhs and rhs must be the same.");
+  }
+  if (lhs.IsPerTensorQuantized()) {
+    if (rhs.IsQuantized() && !output.IsQuantized()) {
+      return absl::FailedPreconditionError(
+          "stablehlo.dot_general: If lhs and rhs are quantized tensors, than "
+          "the output tensor should also be quantized.");
+    } else if (lhs.StorageType() != rhs.StorageType()) {
+      return absl::FailedPreconditionError(
+          "stablehlo.dot_general: If the lhs and rhs are quantized tensors, "
+          "than they should have the same storage type.");
+    } else if (rhs.IsPerTensorQuantized()) {
+      if (!output.IsPerTensorQuantized()) {
+        return absl::FailedPreconditionError(
+            "stablehlo.dot_general: If lhs and rhs are per-tensor quantized "
+            "than output should also be per-tensor quantized.");
+      }
+      if (lhs.quantized_per_tensor_element_type().ExpressedType() ==
+          rhs.quantized_per_tensor_element_type().ExpressedType()) {
+        if (lhs.quantized_per_tensor_element_type().ExpressedType() !=
+            output.quantized_per_tensor_element_type().ExpressedType()) {
+          return absl::FailedPreconditionError(
+              "stablehlo.dot_general: The expressed_type of output tensor must "
+              "be the same as the expressed_type of lhs and rhs tensors.");
+        }
+      }
+      auto check_zero_point_value_is_zero =
+          [](auto zero_point) -> bool {
+        if (std::holds_alternative<I4>(zero_point)) {
+          return std::get<I4>(zero_point) == static_cast<I4>(0);
+        } else if (std::holds_alternative<int8_t>(zero_point)) {
+          return std::get<int8_t>(zero_point) == static_cast<int8_t>(0);
+        } else if (std::holds_alternative<int16_t>(zero_point)) {
+          return std::get<int16_t>(zero_point) == static_cast<int16_t>(0);
+        }
+        return false;
```

Comment:
Does something like this work?
```suggestion
        return std::visit([](const auto v) { return v == static_cast<decltype(v)>(0); }, zero_point);
```

Reply:
this worked.